File: apina-core/src/test/java/fi/evident/apina/spring/SpringAnnotationUtilsTest.java
package fi.evident.apina.spring;

import fi.evident.apina.java.model.JavaAnnotation;
import fi.evident.apina.java.model.type.JavaBasicType;
import org.junit.Test;

import java.util.Optional;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

public class SpringAnnotationUtilsTest {

    private static final JavaBasicType REQUEST_PARAM = new JavaBasicType("org.springframework.web.bind.annotation.RequestParam");
    private static final JavaBasicType REQUEST_MAPPING = new JavaBasicType("org.springframework.web.bind.annotation.RequestMapping");

    @Test
    public void undefinedRequestParamName() {
        JavaAnnotation annotation = new JavaAnnotation(REQUEST_PARAM);
        assertThat(SpringAnnotationUtils.getRequestParamName(annotation), is(Optional.empty()));
    }

    @Test
    public void requestParamNameAsValueAttribute() {
        JavaAnnotation annotation = new JavaAnnotation(REQUEST_PARAM);
        annotation.setAttribute("value", "foo");

        assertThat(SpringAnnotationUtils.getRequestParamName(annotation), is(Optional.of("foo")));
    }

    @Test
    public void requestParamNameAsNameAttribute() {
        JavaAnnotation annotation = new JavaAnnotation(REQUEST_PARAM);
        annotation.setAttribute("name", "foo");

        assertThat(SpringAnnotationUtils.getRequestParamName(annotation), is(Optional.of("foo")));
    }

    @Test
    public void requestMappingPathFromValue() {
        JavaAnnotation annotation = new JavaAnnotation(REQUEST_MAPPING);
        annotation.setAttribute("value", "/foo");

        assertThat(SpringAnnotationUtils.getRequestMappingPath(annotation), is(Optional.of("/foo")));
    }

    @Test
    public void requestMappingPathFromPath() {
        JavaAnnotation annotation = new JavaAnnotation(REQUEST_MAPPING);
        annotation.setAttribute("path", "/foo");

        assertThat(SpringAnnotationUtils.getRequestMappingPath(annotation), is(Optional.of("/foo")));
    }
}

File: apina-core/src/test/kotlin/fi/evident/apina/spring/SpringAnnotationUtilsTest.kt
package fi.evident.apina.spring

import fi.evident.apina.java.model.JavaAnnotation
import fi.evident.apina.java.model.type.JavaType
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class SpringAnnotationUtilsTest {

    @Test
    fun undefinedRequestParamName() {
        val annotation = JavaAnnotation(REQUEST_PARAM)
        assertNull(SpringAnnotationUtils.getRequestParamName(annotation))
    }

    @Test
    fun requestParamNameAsValueAttribute() {
        val annotation = JavaAnnotation(REQUEST_PARAM)
        annotation.setAttribute("value", "foo")

        assertEquals("foo", SpringAnnotationUtils.getRequestParamName(annotation))
    }

    @Test
    fun requestParamNameAsNameAttribute() {
        val annotation = JavaAnnotation(REQUEST_PARAM)
        annotation.setAttribute("name", "foo")

        assertEquals("foo", SpringAnnotationUtils.getRequestParamName(annotation))
    }

    @Test
    fun requestMappingPathFromValue() {
        val annotation = JavaAnnotation(REQUEST_MAPPING)
        annotation.setAttribute("value", "/foo")

        assertEquals("/foo", SpringAnnotationUtils.getRequestMappingPath(annotation))
    }

    @Test
    fun requestMappingPathFromPath() {
        val annotation = JavaAnnotation(REQUEST_MAPPING)
        annotation.setAttribute("path", "/foo")

        assertEquals("/foo", SpringAnnotationUtils.getRequestMappingPath(annotation))
    }

    companion object {

        private val REQUEST_PARAM = JavaType.Basic("org.springframework.web.bind.annotation.RequestParam")
        private val REQUEST_MAPPING = JavaType.Basic("org.springframework.web.bind.annotation.RequestMapping")
    }
}
